modules/kubernetes-addons/crossplane/main.tf

```diff
@@ -17,49 +17,127 @@ module "helm_addon" {
-    iam-role-arn = "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:role/${var.eks_cluster_id}-provider-aws--irsa"
+#--------------------------------------
+# AWS Provider
+#--------------------------------------
+resource "kubectl_manifest" "aws_controller_config" {
+  count = var.aws_provider.enable == true ? 1 : 0
+
+  yaml_body = templatefile("${path.module}/aws-provider/aws-controller-config.yaml", {
+    iam-role-arn = "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:role/${var.eks_cluster_id}-${local.aws_provider_sa}-irsa"
   })
   depends_on = [module.helm_addon]
 }
 
 resource "kubectl_manifest" "aws_provider" {
-  yaml_body = templatefile("${path.module}/aws-provider/provider-aws.yaml", {
-    provider-aws-version = var.crossplane_provider_aws.provider_aws_version
+  count = var.aws_provider.enable == true ? 1 : 0
+
+  yaml_body = templatefile("${path.module}/aws-provider/aws-provider.yaml", {
+    provider-aws-version = var.aws_provider.provider_aws_version
+    aws-provider-name    = local.aws_provider_sa
   })
-  depends_on = [kubectl_manifest.controller_config]
+  depends_on = [kubectl_manifest.aws_controller_config]
 }
 
 module "aws_provider_irsa" {
+  count = var.aws_provider.enable == true ? 1 : 0
+
   source                            = "../../../modules/irsa"
   eks_cluster_id                    = var.eks_cluster_id
   create_kubernetes_namespace       = false
   create_kubernetes_service_account = false
   kubernetes_namespace              = local.namespace
-  kubernetes_service_account        = "provider-aws-*"
-  irsa_iam_policies                 = concat([aws_iam_policy.aws_provider.arn], var.crossplane_provider_aws.additional_irsa_policies)
+  kubernetes_service_account        = "${local.aws_provider_sa}-*"
+  irsa_iam_policies                 = concat([aws_iam_policy.aws_provider[0].arn], var.aws_provider.additional_irsa_policies)
   tags                              = var.tags
 
   depends_on = [kubectl_manifest.aws_provider]
 }
 
 resource "aws_iam_policy" "aws_provider" {
+  count = var.aws_provider.enable == true ? 1 : 0
+
   description = "Crossplane AWS Provider IAM policy"
-  name        = "${var.eks_cluster_id}-aws-provider-irsa"
+  name        = "${var.eks_cluster_id}-${local.aws_provider_sa}-irsa"
   policy      = data.aws_iam_policy_document.s3_policy.json
   tags        = var.tags
 }
 
+resource "kubectl_manifest" "aws_provider_config" {
+  count = var.aws_provider.enable == true ? 1 : 0
+
+  yaml_body = templatefile("${path.module}/aws-provider/aws-provider-config.yaml", {})
+
+  depends_on = [kubectl_manifest.aws_provider, time_sleep.wait_30_seconds_aws]
+}
+
 # Wait for the AWS Provider CRDs to be fully created before initiating aws_provider_config deployment
-resource "time_sleep" "wait_30_seconds" {
-  depends_on = [kubectl_manifest.aws_provider]
+resource "time_sleep" "wait_30_seconds_aws" {
```

Comment:
What is the purpose of the sleep? A random 30 sec sleep seems like a bad idea?

Reply:
I have added a comment above with the justification. kubectl manifest applies the manifest in seconds and comes back with the response as complete however the CRDs are still being created behind the scenes.

This is one of the race condition which i am handling using this time_sleep. This normally takes 10 to 20 seconds to create all the CRDs which is a dependency to the next resource(provider config). The next resource apply fails if these are not ready.  

We can improve this if we find a better solution later .

Reply:
We should follow https://github.com/gavinbunney/terraform-provider-kubectl/issues/41